src/main/scala/sangria/marshalling/ArrayMapBuilder.scala

```diff
@@ -62,4 +63,40 @@ class ArrayMapBuilder[T](keys: Seq[String]) {
 
     builder.result()
   }
+
+  lazy val toIterator: Iterator[(String, T)] = {
+    new Iterator[(String, T)] {
+      var index = -1
+      var nextIndex = 0
+
+      @tailrec def nextIndex(current: Int): Option[Int] = {
+        val next = current + 1
+        if (next >= elements.length) None
+        else if (indexesSet.contains(next)) Some(next)
+        else nextIndex(next)
+      }
+
+      override def hasNext: Boolean = {
+        nextIndex(index) match {
+          case Some(i) ⇒
+            nextIndex = i
+            true
+          case None ⇒
+            false
+        }
+      }
+
+      override def next(): (String, T) = {
+        index = nextIndex
+        elements(nextIndex)
+      }
+    }
+  }
+
+  lazy val toIterable: Iterable[(String, T)] = {
+    val it = toIterator
```

Comment:
I wonder whether this might violate the semantics/contract of `Iterable`. Based on my understanding, in this implementation, `def iterator` will always return the same instance. According to the scaladocs, `iterator` method must:

> Creates a new iterator over all elements contained in this iterable object.

From a practical perspective, I think this code will cause cause an unexpected behaviour:

```scala
val iterable =  builder.toIterable
val iterator1 = iterable.iterator

iterator1.next() // gives element at index 0

val iterator2 = iterable.iterator

iterator2.next() // gives element at index 1
iterator1.next() // gives element at index 2
```

I'm not 100% sure this behaviour  fulfills the contract of `Iterable`. What is your opinion on this?

I also noticed in the scalados of `IterableLike`:

> They also need to provide a method `newBuilder`
> which creates a builder for collections of the same kind.

I wonder whether we should also consider `newBuilder`?

Reply:
I understand your point. We may have to change some `lazy val` in `def`.
I think it's better to extend the tests to be sure the implementation is right.

Reply:
I added tests about this and changed the implementation: https://github.com/sangria-graphql/sangria-marshalling-api/pull/3/commits/28db42b044024f95f0b9286da92ba911d0aa3229

Reply:
> I wonder whether we should also consider newBuilder?

I tried that, but I failed. I got lost in the hierarchy of all traits...